packages/@expo/cli/src/inlineModules/generation.ts

```diff
@@ -0,0 +1,409 @@
+    for (const key of Object.keys(objects.PBXFileSystemSynchronizedRootGroup)) {
+      if (key.endsWith('_comment')) {
+        continue;
+      }
+      if (
+        path.relative(iosFolderPath, path.resolve(projectRoot, dir)) ===
+        objects.PBXFileSystemSynchronizedRootGroup[key].path
+      ) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  const swiftWatchedDirectories =
+    getConfig(projectRoot).exp.inlineModules?.watchedDirectories ?? [];
+  for (const dir of swiftWatchedDirectories) {
+    if (dirEntryExists(dir)) {
+      continue;
+    }
+
+    const newUUID = pbxProject.generateUuid();
+    objects.PBXGroup[mainGroupUUID].children.push({
+      value: newUUID,
+      comment: dir,
+    });
+
+    if (!objects.PBXFileSystemSynchronizedRootGroup) {
+      objects.PBXFileSystemSynchronizedRootGroup = {};
+    }
+
+    objects.PBXFileSystemSynchronizedRootGroup[newUUID] = {
+      isa: 'PBXFileSystemSynchronizedRootGroup',
+      explicitFileTypes: {},
+      explicitFolders: [],
+      name: dir,
+      path: path.relative(iosFolderPath, path.resolve(projectRoot, dir)),
+      sourceTree: 'SOURCE_ROOT',
+    };
+
+    //@ts-ignore
+    objects.PBXFileSystemSynchronizedRootGroup[newUUID + '_comment'] = dir;
+
+    const nativeTargetGroup = objects.PBXNativeTarget[mainTargetUUID];
+    if (!nativeTargetGroup.fileSystemSynchronizedGroups) {
+      nativeTargetGroup.fileSystemSynchronizedGroups = [];
+    }
+    nativeTargetGroup.fileSystemSynchronizedGroups.push({ value: newUUID, comment: dir });
+  }
+
+  fs.writeFileSync(pbxProject.filepath, pbxProject.writeSync());
+}
+
+function getWatchedDirAncestorAbsolutePath(
+  projectRoot: string,
+  filePathAbsolute: string
+): string | null {
+  const watchedDirectories = getConfig(projectRoot).exp.inlineModules?.watchedDirectories ?? [];
+  const realRoot = path.resolve(projectRoot);
+  for (const dir of watchedDirectories) {
+    const dirPathAbsolute = path.resolve(realRoot, dir);
+    if (filePathAbsolute.startsWith(dirPathAbsolute)) {
+      return dirPathAbsolute;
+    }
+  }
+  return null;
+}
+
+function onSourceFileCreated(
+  projectRoot: string,
+  watchedDirRootAbolutePath: string,
+  absoluteFilePath: string,
+  filesWatched?: Set<string>
+): void {
+  const { moduleTypesFilePath, viewTypesFilePath, viewExportPath, moduleExportPath, moduleName } =
+    typesAndLocalModulePathsForFile(projectRoot, watchedDirRootAbolutePath, absoluteFilePath);
+
+  if (filesWatched && fileWatchedWithAnyNativeExtension(absoluteFilePath, filesWatched)) {
+    filesWatched.add(absoluteFilePath);
+    return;
+  }
+  if (filesWatched) {
+    filesWatched.add(absoluteFilePath);
+  }
+
+  fs.mkdirSync(path.dirname(moduleExportPath), { recursive: true });
+  fs.mkdirSync(path.dirname(moduleTypesFilePath), { recursive: true });
+
+  fs.writeFileSync(
+    viewExportPath,
+    `import { requireNativeView } from 'expo';
+export default requireNativeView("${moduleName}");`
+  );
```

Comment:
Creating source files like this can result in race conditions where the file watcher isn't aware of new file during a hot reload / fast refresh. Would be better to use some sort of virtual module or inlined operation instead. Ideally no module.